### ISSUE TITLE
The training process will get stuck after training for one epoch

### ISSUE THREAD

**Issue by leoozy** [IntelLabs/academic-budget-bert#26]
Hi,  @peteriz  seems like there is an issue if deleting the line global_rank = 0. With different worker reading different shard, the total num of iteration for each worker in an epoch is different. So at the end of an epoch, it has synchronization issue and gets stuck. With global_bank=0 the issue disappeared, since the torch data sampler gives each worker the same amount of data. But this has the issue as @sangmichaelxie described, it will skip every 8 files to read.

_Originally posted by @Xinpeng-Wang in https://github.com/IntelLabs/academic-budget-bert/issues/22#issuecomment-1173159490_

As @Xinpeng-Wang said, after you fix the bug by deleting global_bank=0, the code will get stuch after one epoch. Could you please help me to solve the problem?

**Comment by iamsimha** [comment]
This issue is probably related to 

https://github.com/Lightning-AI/lightning/issues/13498

**Comment by iamsimha** [comment]
Update: I raised this issue in the deepspeed repo, turns out Deepspeed doesnot support variable number of batches in each process. To quote their response


"**The common practice with data parallel training is for the gpus to process the same batch size. The hang you are experiencing is unsurprising since imbalance in amount of work done by the gpus will break the tightly coordinated synchronizations. Also, the computation will break since gradient reduction assumes each gpu processes the same amount of data.**"

https://github.com/microsoft/DeepSpeed/issues/2223

I think we will need to refactor dataset iterator, to provide same number of batches for every process.

**Comment by peteriz** [comment]
@iamsimha Thanks for looking into this issue. If you happen to fix the iterator I will gladly review and merge into the main branch.

**Comment by iamsimha** [comment]
Sure @peteriz . I will try this over the weekend.

**Comment by shizhediao** [comment]
> Sure @peteriz . I will try this over the weekend.

Hi did you manage to solve this issue? Thanks!

**Comment by iamsimha** [comment]
Hi, I evaluated two approaches for re-factoring dataset iterator

1. Use shared memory to share imbalanced data between process. However, shared memory has to be allocated at the very beginning of training, since deepspeed takes care of forking processes, we will not be able to allocate shared memory.
2. At the start of the epoch, share additional data between the processes using IPC (torch.gather, torch.scatter etc.). This will not work because of the issue in pytorch dataset class (https://github.com/pytorch/pytorch/issues/77159). We encounter this issue, since we have to allocate cuda tensors in a dataset class.

The only ~~sensible~~( easy :) ) option is to re-balance the shards on disk. This rebalancing should not affect model  performance since, the dataset is already tokenised. I have created a gist to do this. https://gist.github.com/iamsimha/e00b468c082ffe49210d3804e10dc85f

I can create PR if @peteriz agrees with the apporach.

**Comment by shizhediao** [comment]
> Hi, I evaluated two approaches for re-factoring dataset iterator
> 
> 1. Use shared memory to share imbalanced data between process. However, shared memory has to be allocated before creating worker processes. Since deepspeed takes care of forking processes, we will not be able to allocate shared memory.
> 2. At the start of the epoch, share additional data between the processes using IPC (torch.gather, torch.scatter etc.). This will not work because of the issue in pytorch dataset class ([A somewhat cryptic error message (for newcomers) - "Cannot re-initialize CUDA in forked subprocess" - report and suggestion for a possible solution pytorch/pytorch#77159](https://github.com/pytorch/pytorch/issues/77159)). We encounter this issue, since we have to allocate cuda tensors in a dataset class.
> 
> The only ~sensible~( easy :) ) option is to re-balance the shards on disk. This rebalancing should not affect model performance since, the dataset is already tokenised. I have created a gist to do this. https://gist.github.com/iamsimha/e00b468c082ffe49210d3804e10dc85f
> 
> You need to run `python balance_shards.py --dir <path to .hdf5 files>`
> 
> I can create PR if @peteriz agrees with the apporach.

Thanks for your reply and solution!

**Comment by peteriz** [comment]
> Hi, I evaluated two approaches for re-factoring dataset iterator
> 
> 1. Use shared memory to share imbalanced data between process. However, shared memory has to be allocated before creating worker processes. Since deepspeed takes care of forking processes, we will not be able to allocate shared memory.
> 2. At the start of the epoch, share additional data between the processes using IPC (torch.gather, torch.scatter etc.). This will not work because of the issue in pytorch dataset class ([A somewhat cryptic error message (for newcomers) - "Cannot re-initialize CUDA in forked subprocess" - report and suggestion for a possible solution pytorch/pytorch#77159](https://github.com/pytorch/pytorch/issues/77159)). We encounter this issue, since we have to allocate cuda tensors in a dataset class.
> 
> The only ~sensible~( easy :) ) option is to re-balance the shards on disk. This rebalancing should not affect model performance since, the dataset is already tokenised. I have created a gist to do this. https://gist.github.com/iamsimha/e00b468c082ffe49210d3804e10dc85f
> 
> You need to run `python balance_shards.py --dir <path to .hdf5 files>`
> 
> I can create PR if @peteriz agrees with the apporach.

I'm okay with the approach, mainly since its just offline preprocessing, and I also think it will not affect model performance. 
You're welcome to open a PR. Thanks!

**Comment by iamsimha** [comment]
Hi @peteriz , I was going through the code once again. I think global_rank=0, should **not** be deleted (https://github.com/IntelLabs/academic-budget-bert/issues/22#issuecomment-1173159490). 

Even though each process reads the same file, since we are using DistributedSampler https://github.com/IntelLabs/academic-budget-bert/blob/04f6da685acf4dfc47b85b42307e17340e87fde3/pretraining/dataset/distributed_pretraining_dataset.py#L220
each process gets a non-overlapping slice of data and training will proceed without skipping any data.

Please let me know if my understanding is wrong.
